doc/developer/platform/formalism.md

```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+the usual way for join semilattices: for any two times `t1` and `t2` in `T`,
+`t1 <=t t2` iff `join(t1, t2) = t2`. We also define a strict partial order over
+times `<t`, such that `t1 < t2` iff `t1 <= t2` and `t1 != t2`.
+
+`T` contains a minimal time `t0` such that `t0 <=t t` for all `t` in `T`. Times
+are therefore
+[well-founded](https://en.wikipedia.org/wiki/Well-founded_relation).
+
+As a concrete example, we might choose `T` to be the naturals, with `max` as
+our join and `t0 = 0`. The usual senses of `<=` and `<` then serve as
+`<=t` and `<t`, respectively.
+
+A *frontier* is a set of times in `T`. We use frontiers to specify a lower (or
+upper) bound on times in `T`. We say that a time `t` *is later than* or *after*
+a frontier `F` iff there exists some `f` in `F` such that `f <t t`. We write
+this `F <=t t`. Similarly we can speak of a time `t` which is later than or
+equal to F, by which we mean there exists some `f` in `F` such that `f <=t t`:
+we write this `F <=t t`.
+
+For integers, a frontier could be the singleton set `F = {5}`; the set of times
+after `F` are `{6, 7, 8, ...}`.
+
+The latest frontier is the empty set `{}`. No time is later than `{}`.
+
+Throughout this document, "time" generally refers to Materialize time. We say
+"wall-clock time" to describe time in the real world.
+
+## Collection Versions
+
+A *collection version* is a [multiset](https://en.wikipedia.org/wiki/Multiset)
+of data which are all of the same type. Collection versions represent the state
+of a collection at a single instant in time. For example, a collection version
+could store the state of a relation (e.g. table) in SQL, or the state of a
+collection/bucket/keyspace in a key-value store.
+
+The *multiplicity* (or *frequency*) of a piece of data in some collection
+version `c` is, informally speaking, the number of times it appears in `c`. We
+write this as `multiplicity(c, data)`. We can encode a collection version as a
+map of `{data => multiplicity}`.
+
+Multiplicities may be negative: partially ordered delivery of updates to
+collection versions requires commutativity. However, users generally don't see
+collection versions with negative multiplicities.
+
+The *empty* collection version `cv0` is an empty multiset.
+
+## Time-varying collections
+
+A *time-varying collection* (TVC) represents the entire history of a collection
+which changes over time. At any time, it has an effective value: a collection
+version. We model a TVC as a map of `{time => collection-version}` for all
+possible times.
+
+A *read* of a TVC `tvc` at time `t` means the collection version stored in
+`tvc` for `t`. We write this as `read(tvc, t) => collection-version`.
+
+Every TVC begins with the empty collection version: `read(tvc, t0) = cv0`.
+
+## Partial time-varying collections
+
+At any wall-clock time a TVC is (in general) unknowable. Some of its past
+changes have been forgotten, some information has yet to arrive, and its future
```

Comment:
No strong feels about this. I thought it might be useful color, but I think I need an emoji to indicate :bonus: or something like that. 🍖 is the closest github offers me.